### PR TITLE
Add gtk-primary-selection protocol

### PR DESCRIPTION
**Pull request by emersion**
Test plan: open two different GNOME apps, select some text on one of them, focus the other, and press the middle mouse button. The text should be pasted.

### CODE REVIEW

**Comment by SirCmpwn** [comment]
This should be synchronized with the xwayland primary selection when it's ready. Also could use a test plan.

**Comment by emersion** [comment]
I don't feel like making xwayland support in this PR, I'll do that in another one. I'll add a test plan when ready.

**Comment by SirCmpwn** [comment]
:+1:

**Comment by emersion** [comment]
This PR is ready for review.

File: types/wlr_primary_selection.c [hunk 1]
```diff
@@ -1,10 +1,118 @@
...
 #include <stdlib.h>
+#include <unistd.h>
 #include <string.h>
 #include <gtk-primary-selection-protocol.h>
 #include <wlr/types/wlr_primary_selection.h>
+#include <wlr/types/wlr_seat.h>
 #include <wlr/util/log.h>
 
+static void client_source_accept(struct wlr_primary_selection_source *source,
+		uint32_t serial, const char *mime_type) {}
```
Comment (SirCmpwn) [inline types/wlr_primary_selection.c]: Code style here is pretty bad

Comment (SirCmpwn) [reply types/wlr_primary_selection.c]: Prefer:

```
... {
    // This space intentionally left blank
}
```

**Comment by SirCmpwn** [comment]
Was looking at an old diff. This LGTM.

File: types/wlr_primary_selection.c [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+	}
+
+	manager->display_destroy.notify = handle_display_destroy;
+	wl_display_add_destroy_listener(display, &manager->display_destroy);
+
+	return manager;
+}
+
+void wlr_primary_selection_device_manager_destroy(
+		struct wlr_primary_selection_device_manager *manager) {
```
Comment (acrisci) [inline types/wlr_primary_selection.c]: need to clean up `display_destroy` listener.

File: types/wlr_primary_selection.c [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+		uint32_t serial) {
+	struct wlr_primary_selection_source *source = NULL;
+	if (source_resource != NULL) {
+		source = wl_resource_get_user_data(source_resource);
+	}
+
+	struct wlr_seat_client *seat_client =
+		wl_resource_get_user_data(resource);
+
+	// TODO: store serial and check against incoming serial here
```
Comment (acrisci) [inline types/wlr_primary_selection.c]: How could the serial be validated? Serial validation is currently thought to be a responsibility of rootston. Should we add the ability for the seat itself to validate serials and use this approach in rootston as well?

File: types/wlr_primary_selection.c [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+
+static const struct gtk_primary_selection_source_interface source_impl = {
+	.offer = source_handle_offer,
+	.destroy = source_handle_destroy,
+};
+
+static void source_resource_handle_destroy( struct wl_resource *resource) {
+	struct wlr_primary_selection_source *source =
+		wl_resource_get_user_data(resource);
+	wl_signal_emit(&source->events.destroy, source);
```
Comment (acrisci) [inline types/wlr_primary_selection.c]: Need to free the mime types here?

File: types/wlr_primary_selection.c (hunk 3 above)
Comment (acrisci) [reply types/wlr_primary_selection.c]: I think #511 is the right way to do this.

**Comment by emersion** [comment]
Fixed @acrisci's comments

**Comment by acrisci** [comment]
👍

**Pull request merged**